pulsar-client-cpp/lib/c/c_Consumer.cc

```diff
@@ -60,6 +60,20 @@ pulsar_result pulsar_consumer_receive_with_timeout(pulsar_consumer_t *consumer,
     return (pulsar_result)res;
 }
 
+static void handle_receive_callback(pulsar::Result result, pulsar::Message message,
+                                    pulsar_receive_callback callback, void *ctx) {
+    if (callback) {
+        pulsar_message_t *msg = new pulsar_message_t;
+        msg->message = message;
+        callback((pulsar_result)result, msg, ctx);
```

Comment:
I think it's better to delete `pulsar_message_t` here and don't leave users to call `pulsar_message_free`. We should encourage users to call functions with `pulsar_message_` prefix in the callback and pass the results to `ctx`.

For a C user, it's more natural to write code like:

```c++
struct receive_ctx {
    pulsar_result result;
    const char *value;
    uint32_t length;
};

static void receive_callback(pulsar_result result, pulsar_message_t *msg, void *ctx) {
    struct receive_ctx *receive_ctx = (struct receive_ctx *)ctx;
    receive_ctx->result = result;
    if (result == pulsar_result_Ok) {
        receive_ctx->value = (const char *)pulsar_message_get_data(msg);
        receive_ctx->length = pulsar_message_get_length(msg);
    } else {
        receive_ctx->value = NULL;
        receive_ctx->length = 0;
    }
}
```

But for the current design, he must save the `msg` in the `ctx` or a global variable, then call `pulsar_message_free` after that.

Reply:
In my understanding, other async functions are also defined by the current approach.
https://github.com/apache/pulsar/blob/a7f1a5657782ff7ac91a89afdd49568d74d111f5/pulsar-client-cpp/lib/c/c_Producer.cc#L39-L55

I think it is better not to change semantics in the same C API.

Reply:
It makes sense to me. I think we should note it in the docs, or add it to examples.

Reply:
Okay. I'll add an example about async consume in this PR. Also, I'll create an issue about other async functions.